### PR TITLE
dbplyr 2.0.0 backend

### LINKED ISSUE #132: <Pool> uses an old dbplyr interface warning (and e.g. error when using stringr)
Due, I believe, to `dbplyr` latest update on CRAN I am receiving a warning and errors in my scripts (as well as in [a CRAN package](https://cran.r-project.org/web/checks/check_results_tidywikidatar.html) I maintain) that relies on `pool`. 

The warning seems to appear even with common filters. The error appears in filters that used to work fine until the `dbplyr` update, but that perhaps did not have a fully compliant syntax to begin with. 

Reprex below should show the issue. Latest version on CRAN and current development version from the repository show the same behaviour.

``` r
library("dplyr", warn.conflicts = FALSE)

months_df <- tibble::tibble(abb = head(month.abb),
                          name = head(month.name))

months_df
#> # A tibble: 6 × 2
#>   abb   name    
#>   <chr> <chr>   
#> 1 Jan   January 
#> 2 Feb   February
#> 3 Mar   March   
#> 4 Apr   April   
#> 5 May   May     
#> 6 Jun   June

db <- pool::dbPool(
  drv = RSQLite::SQLite(),
  dbname = paste0(tempfile(), "sqlite")
)

pool::dbWriteTable(db,
                   name = "months",
                   value = months_df,
                   append = TRUE
)

pool::dbReadTable(conn = db,
                  name = "months")
#>   abb     name
#> 1 Jan  January
#> 2 Feb February
#> 3 Mar    March
#> 4 Apr    April
#> 5 May      May
#> 6 Jun     June

# just a warning
dplyr::tbl(src = db, "months") %>%
  dplyr::filter(.data$abb == "Apr")
#> Warning: <Pool> uses an old dbplyr interface
#> ℹ Please install a newer version of the package or contact the maintainer
#> This warning is displayed once every 8 hours.
#> # Source:   SQL [1 x 2]
#> # Database: sqlite 3.38.5 [/tmp/Rtmp3yO10e/filea7e172c2d5eesqlite]
#>   abb   name 
#>   <chr> <chr>
#> 1 Apr   April

dplyr::tbl(src = db, "months") %>%
  dplyr::filter(.data$abb %in% c("Feb", "Apr"))
#> # Source:   SQL [2 x 2]
#> # Database: sqlite 3.38.5 [/tmp/Rtmp3yO10e/filea7e172c2d5eesqlite]
#>   abb   name    
#>   <chr> <chr>   
#> 1 Feb   February
#> 2 Apr   April


## error
dplyr::tbl(src = db, "months") %>%
  dplyr::filter(.data$abb %in% stringr::str_to_title(c("feb", "apr")))
#> Error: no such table: INITCAP

dplyr::tbl(src = db, "months") %>%
  dplyr::filter(.data$abb %in% stringr::str_to_upper(c("feb", "apr")))
#> Error: no such table: UPPER

dplyr::tbl(src = db, "months") %>%
  dplyr::filter(.data$abb %in% stringr::str_c(c("Feb", "Apr")))
#> Error: no such table: CONCAT_WS

## but can prevent with !!

dplyr::tbl(src = db, "months") %>%
  dplyr::filter(.data$abb %in% !!stringr::str_to_title(c("feb", "apr")))
#> # Source:   SQL [2 x 2]
#> # Database: sqlite 3.38.5 [/tmp/Rtmp3yO10e/filea7e172c2d5eesqlite]
#>   abb   name    
#>   <chr> <chr>   
#> 1 Feb   February
#> 2 Apr   April
```

<sup>Created on 2022-06-06 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value
#>  version  R version 4.1.3 (2022-03-10)
#>  os       Fedora Linux 35 (Workstation Edition)
#>  system   x86_64, linux-gnu
#>  ui       X11
#>  language (EN)
#>  collate  en_IE.UTF-8
#>  ctype    en_IE.UTF-8
#>  tz       Europe/Vienna
#>  date     2022-06-06
#>  pandoc   2.14.0.3 @ /usr/libexec/rstudio/bin/pandoc/ (via rmarkdown)
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version    date (UTC) lib source
#>  assertthat    0.2.1      2019-03-21 [1] CRAN (R 4.1.2)
#>  bit           4.0.4      2020-08-04 [1] CRAN (R 4.1.1)
#>  bit64         4.0.5      2020-08-30 [1] CRAN (R 4.1.1)
#>  blob          1.2.3      2022-04-10 [1] CRAN (R 4.1.3)
#>  cachem        1.0.6      2021-08-19 [1] CRAN (R 4.1.1)
#>  cli           3.3.0      2022-04-25 [1] CRAN (R 4.1.3)
#>  crayon        1.5.1      2022-03-26 [1] CRAN (R 4.1.2)
#>  DBI           1.1.2      2021-12-20 [1] CRAN (R 4.1.2)
#>  dbplyr        2.2.0      2022-06-05 [1] CRAN (R 4.1.3)
#>  digest        0.6.29     2021-12-01 [1] CRAN (R 4.1.2)
#>  dplyr       * 1.0.9      2022-04-28 [1] CRAN (R 4.1.3)
#>  ellipsis      0.3.2      2021-04-29 [2] CRAN (R 4.1.0)
#>  evaluate      0.15       2022-02-18 [1] CRAN (R 4.1.2)
#>  fansi         1.0.3      2022-03-24 [1] CRAN (R 4.1.2)
#>  fastmap       1.1.0      2021-01-25 [2] CRAN (R 4.1.0)
#>  fs            1.5.2.9000 2022-05-31 [1] Github (r-lib/fs@e7d98c4)
#>  generics      0.1.2      2022-01-31 [1] CRAN (R 4.1.2)
#>  glue          1.6.2      2022-02-24 [1] CRAN (R 4.1.2)
#>  highr         0.9        2021-04-16 [3] CRAN (R 4.1.0)
#>  htmltools     0.5.2      2021-08-25 [1] CRAN (R 4.1.1)
#>  knitr         1.39       2022-04-26 [1] CRAN (R 4.1.3)
#>  later         1.3.0      2021-08-18 [1] CRAN (R 4.1.2)
#>  lifecycle     1.0.1      2021-09-24 [1] CRAN (R 4.1.1)
#>  magrittr      2.0.3      2022-03-30 [1] CRAN (R 4.1.3)
#>  memoise       2.0.1      2021-11-26 [1] CRAN (R 4.1.2)
#>  pillar        1.7.0      2022-02-01 [1] CRAN (R 4.1.2)
#>  pkgconfig     2.0.3      2019-09-22 [1] CRAN (R 4.1.1)
#>  pool          0.1.6      2021-01-14 [1] CRAN (R 4.1.2)
#>  purrr         0.3.4      2020-04-17 [1] CRAN (R 4.1.2)
#>  R.cache       0.15.0     2021-04-30 [3] CRAN (R 4.1.0)
#>  R.methodsS3   1.8.1      2020-08-26 [3] CRAN (R 4.1.0)
#>  R.oo          1.24.0     2020-08-26 [3] CRAN (R 4.1.0)
#>  R.utils       2.11.0     2021-09-26 [1] CRAN (R 4.1.2)
#>  R6            2.5.1      2021-08-19 [1] CRAN (R 4.1.1)
#>  Rcpp          1.0.8.3    2022-03-17 [1] CRAN (R 4.1.2)
#>  reprex        2.0.1      2021-08-05 [1] CRAN (R 4.1.1)
#>  rlang         1.0.2      2022-03-04 [1] CRAN (R 4.1.2)
#>  rmarkdown     2.14       2022-04-25 [1] CRAN (R 4.1.3)
#>  RSQLite       2.2.14     2022-05-07 [1] CRAN (R 4.1.3)
#>  rstudioapi    0.13       2020-11-12 [1] CRAN (R 4.1.2)
#>  sessioninfo   1.2.2      2021-12-06 [1] CRAN (R 4.1.2)
#>  stringi       1.7.6      2021-11-29 [1] CRAN (R 4.1.2)
#>  stringr       1.4.0      2019-02-10 [1] CRAN (R 4.1.2)
#>  styler        1.7.0      2022-03-13 [1] CRAN (R 4.1.2)
#>  tibble        3.1.7      2022-05-03 [1] CRAN (R 4.1.3)
#>  tidyselect    1.1.2      2022-02-21 [1] CRAN (R 4.1.2)
#>  utf8          1.2.2      2021-07-24 [1] CRAN (R 4.1.1)
#>  vctrs         0.4.1      2022-04-13 [1] CRAN (R 4.1.3)
#>  withr         2.5.0      2022-03-03 [1] CRAN (R 4.1.2)
#>  xfun          0.31       2022-05-10 [1] CRAN (R 4.1.3)
#>  yaml          2.3.5      2022-02-21 [1] CRAN (R 4.1.2)
#> 
#>  [1] /home/g/R/x86_64-redhat-linux-gnu-library/4.1
#>  [2] /usr/lib64/R/library
#>  [3] /usr/share/R/library
#> 
#> ──────────────────────────────────────────────────────────────────────────────
```

</details>

### PR DESCRIPTION
**Pull request by hadley**
This replaces custom dbplyr methods with code generated methods, and updates the registered methods for the dbplyr 2.0.0 interface (Fixes #132). The number of methods that need to be registered is now much smaller because we no longer (if we ever?) need to register methods for SQL translation — these are all handled internally by dbplyr after the connection has been checked out from the pool. I've extended the tests a little bit to verify that this true for more cases.

@lionel- can you please take a look at my method registration and dbplyr wrapping code?

### CODE REVIEW

File: R/dbplyr.R [hunk 1]
```diff
@@ -76,257 +37,140 @@ tbl.Pool <- function(src, from, ..., vars = NULL, con = NULL) {
...
-  dplyr::sql_subquery(db_con, from = from, name = name, ...)
-}
-
-#' @rdname dplyr-db-methods
-sql_translate_env.Pool <- function(con) {
-  db_con <- poolCheckout(con)
-  on.exit(poolReturn(db_con))
-  dplyr::check_dbplyr()
-  dplyr::sql_translate_env(db_con)
+    new_function(args, body, env = ns_env("pool"))
```
Comment (lionel-) [inline R/dbplyr.R:112]: If possible, I think it's better not to override the environment. This way the code above can be read in (lexical) context.

**Review by lionel- — APPROVED** [review]
Looks good!

I think this could be simplified a bit though. The `lazy` variant of `s3_register()` is a bit of a distraction and adds complexity. IIUC, it's necessary because the generated functions need to inspect the formals of functions in dbplyr, and we don't want to cause an early load of dbplyr.

I think this could be simplified by calling `s3_register()` within `rlang::on_package_load()`:

```r
on_package_load("dbplyr", {
  # Now `db_wrap()` returns a method instead of a thunk to a method
  register_s3_method("dbplyr", "foo", "Pool", fn = db_wrap("foo"))
})
```

I also tried to find a way of avoiding looking up formals and passing arguments (i.e. use more of a functional approach than a metaprogramming one), but this doesn't seem possible given how the `temporary` argument is set up. It would be helpful to add some comments about why this argument handling is necessary.

Possibly dbplyr should handle this in the generics that have a `temporary` method. We'd add a `dbplyr_allows_temporary()` generic that defaults to `TRUE`, which pool would override to `FALSE`, and this would be checked by the generics instead of individual methods. Then we wouldn't need as much metaprogramming here.

**Comment by hadley** [comment]
@lionel- dbplyr does actually have a `dbplyr::db_table_temporary()` (for another purpose) which I thought we could (ab)use here. Unfortunately is it gets called fairly deep in dbplyr internals, so the pool has already been converted to a regular connection.

Thanks for the `on_package_load()` pointer; I somehow got stuck on the idea of doing everything in `.onLoad()` 😬

File: R/dbplyr.R [hunk 2]
```diff
@@ -76,257 +37,140 @@ tbl.Pool <- function(src, from, ..., vars = NULL, con = NULL) {
...
 
-#' @rdname dplyr-db-methods
-db_query_fields.Pool <- function(con, sql, ...) {
-  db_con <- poolCheckout(con)
-  on.exit(poolReturn(db_con))
-  dplyr::check_dbplyr()
-  dplyr::db_query_fields(db_con, sql = sql, ...)
-}
+      db_con <- poolCheckout(con)
+      on.exit(poolReturn(db_con))
```
[the remark this answers is not in the document]
Comment (hadley) [reply R/dbplyr.R:107]: Yeah, because the methods are now dbplyr generics, not dplyr generics, so dbplyr must be loaded.

**Pull request merged**